src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs

```diff
@@ -96,10 +122,20 @@ public class HttpRetryHandler : IHttpRetryHandler
                         if (response.Content != null)
                         {
                             var networkStream = await response.Content.ReadAsStreamAsync();
-                            var newContent = new DownloadTimeoutStreamContent(
+                            var timeoutStream = new DownloadTimeoutStream(requestUri, networkStream, request.DownloadTimeout);
+                            var diagnosticsInfo = new ProtocolDiagnosticEvent(DateTime.MinValue,
+                                source,
                                 requestUri,
-                                networkStream,
-                                request.DownloadTimeout);
+                                headerStopwatch?.Elapsed,
+                                TimeSpan.MinValue,
+                                (int)response.StatusCode,
+                                bytes: 0,
+                                isSuccess: true,
+                                isRetry: request.IsRetry || tries > 1,
+                                isCancelled: false);
+                            var diagnosticsStream = new ProtocolDiagnosticsStream(timeoutStream, diagnosticsInfo, stopwatch);
```

Comment:
Why is need to pass the stopwatch? Would the elapsed milliseconds be enough ?  I did not see this stopwatch being closed.

Reply:
At this point, the response body hasn't been downloaded. This means the value will be basically the same as the header stopwatch, and the larger the response body length (the longer the download takes), the more inaccurate the reported time would be.

`ProtocolDiagnosticsStream` gets the elapsed time from the stopwatch's `Elaposed` property. `Stopwatch` isn't disposable, and if you look at the implementation, it's effectively just capturing a timestamp when you call `.Start()`, and calculating the difference when you call `.Elapsed` or `.Stop()`. That's the reason `ProtocolDiagnosticsStream` doesn't call `.Stop()`, it doesn't really do anything.